### ISSUE TITLE
Proposal for new methods in VaultObjectOperations and VaultObjectPropertyOperations wrapper

### ISSUE THREAD

**Issue by dmusic** [M-Files/MFilesSamplesAndLibraries#12]
Hi there, 

Thank you sharing this client implementation with a community. 

We have used it as a base for implementation of our own document uploader application setting up random values for metadata of internally configured classes. 
As we have setup our own taxonomy of object types and document types, we needed to read whole structure of our vault including object types (and instances), classes, all properties etc. and to prepare cached values we could use for document class and property filling in our application.
For that we needed additional methods to read single properties and as well latest object versions for our own object types.

We have add following:

In MFWSVaultObjectOperations.cs:

```
/// <summary>
/// Retrieves the latest versions of all object type instances.
/// </summary>
/// <param name="objTypeId">The id of the object type.</param>
/// <param name="token">A cancellation token for the request.</param>
/// <returns>The list of ObjectVersions, if it is visible for user.</returns>
public async Task<Results<ObjectVersion>> GetLatestObjectVersionAsync(int objTypeId, CancellationToken token = default) 
		{
			// Create the request
			var resource = $"/REST/objects/{objTypeId}";
			var request = new RestRequest(resource);

			// Make the request and get the response.
			var response = await this.MFWSClient.Get<Results<ObjectVersion>>(request, token).ConfigureAwait(false);

			// Return the object data.
			return response.Data;
		}

```
and in MFWSVaultObjectPropertyOperations.cs:

```
/// <summary>
		/// Retrieves the specific property of a single object.
		/// </summary>
		/// <param name="objVer">The object to retrieve the properties of.</param>
		/// <param name="propertyDefId">Property definition id it is searched for.</param>
		/// <param name="token">A cancellation token for the request.</param>
		/// <returns>A collection of property values for the supplied object.</returns>
		public PropertyValue GetProperty(ObjVer objVer, int propertyDefId, CancellationToken token = default) 
		{
			return this.GetPropertyAsync(objVer, propertyDefId, token)
					   .ConfigureAwait(false)
					   .GetAwaiter()
					   .GetResult();
		}

		/// <summary>
		/// Retrieves the specific property of a single object.
		/// </summary>
		/// <param name="objVer">The object to retrieve the properties of.</param>
		/// <param name="propertyDefId">Property definition id it is searched for.</param>
		/// <param name="token">A cancellation token for the request.</param>
		/// <returns>A collection of property values for the supplied object.</returns>
		public async Task<PropertyValue> GetPropertyAsync(ObjVer objVer, int propertyDefId, CancellationToken token = default)
		{
			// Sanity.
			if (null == objVer)
				throw new ArgumentNullException(nameof(objVer));

			// Extract the URI elements.
			objVer.GetUriParameters(out int objectTypeId, out string objectId, out string objectVersionId);

			// Create the request.
			// Ignore version and always take the latest version of country object (-1)
			string resource = $"/REST/objects/{objectTypeId}/{objectId}/-1/properties/{propertyDefId}";

			var request = new RestRequest(resource);

			// Make the request and get the response.
			var response = await MFWSClient.Get<PropertyValue>(request, token).ConfigureAwait(false);

			// Return the data.
			return response.Data;
		}
```

We experienced also a small bug due to some dummy documents that had no extension, so we fixed that as well (MFWSVaultObjectFileOperations.cs, line 424):

`uploadInfo.Extension = (file.Extension == "" ? file.Extension : file.Extension.Substring(1)); // Remove the dot if existing.`

**Comment by CraigHawker** [comment]
Hi @dmusic,

Thanks for commenting!

If you wanted to wrap these changes into a merge request then we'd definitely look at pulling them in.  A few small points I would mention that I'd probably ask you to factor in, though:

- Be aware that your `GetLatestObjectVersionAsync` method is probably badly named.  It returns objects of a specific type (yes, the latest versions), but it's not the "LatestObjectVersion".

- The idea of the library was to mirror - where possible - the COM API structure.  In the case of the `GetLatestObjectVersion` method, there's no real equivalent.  The equivalent is to execute a search for all non-deleted objects of the given type.  _Because of this I am unsure whether it's a good candidate for inclusion._

- Speaking of which, do note that `GetLatestObjectVersionAsync` does not necessarily return all objects of a given type; the search results will be limited if you start getting large numbers of objects in your vault.

- The version number in `GetPropertyAsync` is often important.  If you want an overload that gets the latest version then see how something like https://github.com/M-Files/MFilesSamplesAndLibraries/blob/39aac5abf033de197a2bc5df754d8bca8bb3485f/Libraries/MFaaP.MFWSClient/MFWSVaultObjectPropertyOperations.cs#L161 does it (additional overload without the version data).  But if a user wants to provide an ObjVer then the method should most definitely use that version and not throw it away (as it can lead to incorrect results).

- In `GetPropertyAsync` (for consistency, and in line with the documentation) the version number should be the string "latest" if you want the latest version (not -1).  You are already calling ObjVer.GetUriParameters (https://github.com/M-Files/MFilesSamplesAndLibraries/blob/8dbefeeba4e106efe7142eb5b9e1a07ebb9bd5a3/Libraries/MFaaP.MFWSClient/MFWSStructs.cs#L1695) so the string returned for the version should be used.

- In `GetPropertyAsync` (again, solely to be consistent with the rest of the code base), your endpoint should end in ".aspx" ($"/REST/objects/{objectTypeId}/{objectId}/-1/properties/{propertyDefId}.aspx").  Whilst uncommon these days, some older servers are misconfigured to incorrectly handle the extensionless endpoints and will not return the correct data.

Regards,

Craig.

**Comment by CraigHawker** [comment]
Hi dmusic,

This project has been migrated to its own repository.  If you still consider these items worth looking at, and would like to submit a merge request, please open another issue in the [new repository](https://github.com/M-Files/Libraries.MFWSClient).

Regards,

Craig.

**Comment by dmusic** [comment]
Hi Craig,
Sorry, never have time to come back to the topic. But now reviewing it again. I have just found out that a kind of ineffectiveness in api when searching for linked object types. I have described it in a forum. 
I would like to provide others easy to use API which wraps complexity of versioning in M-Files but also to provide understandable mapping between business IDs and m-files object IDs. I have raised this in forum.

Best regards,
Dejan
